packages/core/core-flows/src/cart/workflows/refresh-cart-items.ts

```diff
@@ -0,0 +1,131 @@
+
+export const refreshCartItemsWorkflowId = "refresh-cart-items"
+/**
+ * This workflow refreshes a cart's items
+ */
+export const refreshCartItemsWorkflow = createWorkflow(
+  refreshCartItemsWorkflowId,
+  (
+    input: WorkflowData<{
+      cart_id: string
+      promo_codes?: string[]
+    }>
+  ) => {
+    const cart = useRemoteQueryStep({
+      entry_point: "cart",
+      fields: cartFieldsForRefreshSteps,
+      variables: { id: input.cart_id },
+      list: false,
+    })
+
+    const variantIds = transform({ cart }, (data) => {
+      return (data.cart.items ?? []).map((i) => i.variant_id)
+    })
+
+    const pricingContext = transform(
+      { cart },
+      ({ cart: { currency_code, region_id, customer_id } }) => {
+        return {
+          currency_code,
+          region_id,
+          customer_id,
+        }
+      }
+    )
+
+    const variants = useRemoteQueryStep({
+      entry_point: "variants",
+      fields: productVariantsFields,
+      variables: {
+        id: variantIds,
+        calculated_price: {
+          context: pricingContext,
+        },
+      },
+      throw_if_key_not_found: true,
+    }).config({ name: "fetch-variants" })
+
+    validateVariantPricesStep({ variants })
+
+    const lineItems = transform({ cart, variants }, ({ cart, variants }) => {
+      const items = cart.items.map((item) => {
+        const variant = variants.find((v) => v.id === item.variant_id)!
+
+        const preparedItem = prepareLineItemData({
+          variant: variant,
+          unitPrice: variant.calculated_price.calculated_amount,
+          isTaxInclusive:
+            variant.calculated_price.is_calculated_price_tax_inclusive,
+          quantity: item.quantity,
+          metadata: item.metadata,
+          cartId: cart.id,
+        })
+
+        return {
+          selector: { id: item.id },
+          data: preparedItem,
+        }
+      })
+
+      return items
+    })
+
+    const items = updateLineItemsStep({
+      id: cart.id,
+      items: lineItems,
+    })
+
+    const refetchedCart = useRemoteQueryStep({
+      entry_point: "cart",
+      fields: cartFieldsForRefreshSteps,
+      variables: { id: cart.id },
+      list: false,
+    }).config({ name: "refetch–cart" })
+
+    refreshCartShippingMethodsStep({ cart: refetchedCart })
+
+    updateTaxLinesWorkflow.runAsStep({
+      input: { cart_id: cart.id, items },
+    })
+
+    const cartPromoCodes = transform({ cart, input }, ({ cart, input }) => {
+      if (isDefined(input.promo_codes)) {
+        return input.promo_codes
+      } else {
+        return cart.promotions.map((p) => p.code)
+      }
+    })
+
+    updateCartPromotionsWorkflow.runAsStep({
+      input: {
+        cart_id: cart.id,
+        promo_codes: cartPromoCodes,
+        action: PromotionActions.REPLACE,
+      },
+    })
+
+    refreshPaymentCollectionForCartWorkflow.runAsStep({
+      input: { cart_id: cart.id },
+    })
+
+    return new WorkflowResponse(refetchedCart)
```

Comment:
suggestion: should we really return the cart here knowing that the totals etc eventully wont be correct at this point

Reply:
it is used for the hook right? but in that case the hook will received a non fully updated cart, we just need to be clear about it right?

Reply:
Yes, but I think that approach is wrong. The hooks should ideally be refetching the cart and we should only pass IDs into the hooks and never work with an entire deep object that always has the potential to mutate in multiple ways or go stale if async.

We've already fixed so many bugs internally because of issues on relying on an object passed to a step that at some point changed and broke all the subsequent workflows.